TST: Reproducer for SkylakeX Issue gh-13401

tylerjereddy opened the pull request:
CI-based Intel SDE (SkylakeX emulation) reproducer for #13401.

Many thanks to @isuruf and @martin-frbg for help upstream. For gory details: https://github.com/xianyi/OpenBLAS/issues/2108

This only works for Mac because our MacPython (ecosystem) Linux OpenBLAS was built via `maylinux1` containers where the gcc version is not high enough to empower the AVX switch that triggers the problem in OpenBLAS. So, from the standpoint of our recently-released wheels, Linux should be ok for this matter, consistent with pip install reports in the original issue only failing on Mac. 

`conda` is another story--there you can induce the failure on Linux by using recent OpenBLAS built outside manylinux1 constraints as @isuruf observed today.

This is really just the beginning of something that is going somewhere (?) and perhaps not in NumPy repo proper. When BLIS does this they go through a for loop of emulation archs: https://github.com/flame/blis/blob/master/travis/do_sde.sh

Perhaps the short-term objective here is simply a public CI run that fails because of the reported OpenBLAS issue, and provides a platform for us to try the latest OpenBLAS binary (build in our MacPython ecosystem) to see if it helps for that specific issue.

Plans beyond that are perhaps contentious, as discussed in #13401, but maybe seeing this will help drive that discussion in one direction or another. One problem upstream is that there may not be a maintainer with sufficient OpenBLAS repo access rights to set up new CI for Azure, etc. But they are open to the idea, it seems.

tylerjereddy committed:
TST: reproducer for Issue 13401

* draft of new test and Azure CI job
to detect the OpenBLAS 0.3.5.dev issue
on Mac versions of NumPy 1.16.3 wheel

* use Intel SDE to emulate / force SkylakeX
usage by OpenBLAS in Azure job

tylerjereddy commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -2002,3 +2002,10 @@ def test_unsupported_commontype():
     with assert_raises_regex(TypeError, "unsupported in linalg"):
         linalg.cholesky(arr)
+
+def test_issue_13401():
```
Only this & one other linalg test fail with SkylakeX emulation, even in full mode, which surprised me.

tylerjereddy commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -2002,3 +2002,10 @@ def test_unsupported_commontype():
+    X = np.random.randn(100, 100) * 10
+    u, s, v = np.linalg.svd(X)
+    assert np.mean(X - u @ np.diag(s) @ v) < 1.0
+    assert np.std(X - u @ np.diag(s) @ v) < 1.0
```
these thresholds probably need to be tighter if we were to seriously consider this test--I'll let the linalg gurus comment on that

rgommers replied:
Just see when it fails on a regular run, and then set the threshold one or two orders of magnitude less tight. I'd expect 1e-12 or so to be reasonable here.

isuruf replied:
You also need to do absolute value for `np.mean`.

tylerjereddy replied:
made these two revisions now

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -223,3 +223,54 @@ jobs:
+    displayName: 'Build / Install NumPy'
+  - script: |
+      pushd ..
+      python -c "import numpy, ctypes, os; dll = ctypes.CDLL(numpy.core._multiarray_umath.__file__); get_config = dll.openblas_get_config; get_config.restype=ctypes.c_char_p; res = get_config(); print('OpenBLAS get_config returned', str(res)); assert b'OpenBLAS 0.3.5.dev' in res"
```
reviewers should notice how the reported architecture for OpenBLAS here differs from the string reported under emulation below

martin-frbg replied:
Wait, if you see SandyBridge reported by OpenBLAS you are nowhere near testing for any AVX512 issues (nor AVX2 for that matter)...

isuruf replied:
Sandybridge was before SDE emulation. SkylakeX is after emulation.

tylerjereddy commented:
The expected / desired behavior, if / when I try the newer OpenBLAS binary on our rackspace server in this PR, is likely just that the SkylakeX instructions are not used and there's a fallback to an older arch along with test passing, and NOT that the AVX instruction stuff is fixed, since that appears to be the case for the temporary fix upstream.

I also don't know if `test_sdot_bug_8577` would always fail under emulation, or if it is actually failing because of recent / pertinent OpenBLAS issue.

tylerjereddy committed:
MAINT: revise PR 13466

* use stricter threshold for test_issue_13401
and take absolute value of mean result

* skip a failing sdot bug test that is not
currently compatible with Intel SDE emulation

* switch to "full" linalg test suite for Intel SDE
Azure CI job

tylerjereddy commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1828,6 +1828,7 @@ def test_xerbla_override():
             pytest.skip('Numpy xerbla not linked in.')
 
 
+@pytest.mark.skip(reason='gh-13478')
```
This is unfortunate, but I couldn't manage to get i.e., an environment variable to pass through the SDE for selective skipping in the CI job.

It would also be useful if I could use `runtests.py -s "linalg" -- -k "not test_sdot_bug_8577"`, but I don't think mixing and matching test selection between `runtests.py` and the args fed to `--` (pytest) is fool-proof. 

A further complication is the potential special meaning of `--` after the emulation command, and not sure how well emulation works with pytest-xdist if we want to use many cores...

tylerjereddy reviewed (commented):
Revised with another commit based on feedback & what I thought was still needed here.

If this works as intended, CI should be all green except for 1 failure in the SDE Azure job for the issue we are concerned with.

The next step, if that happens, would be to bump the version of MacPython OpenBLAS used in CI (probably for all matrix entries, but especially for the SDE). If that solves the issue, great, otherwise the are already more patches related to SkylakeX stuff upstream & also a C-level reproducer, though not in their CI yet I don't think.

tylerjereddy committed:
TST, MAINT: bump Azure OpenBLAS

* bump the version of MacPython ecosystem
OpenBLAS used in Azure CI to see if the SkylakeX
issue is resolved for Intel SDE job, while expecting
all other Azure entries to also pass CI with latest
ecosystem OpenBLAS

tylerjereddy commented:
Since the CI did indeed produce only the single failure related to #13401 under emulation, I've now pushed the commit that bumps OpenBLAS to v0.3.7.dev in all Azure CI entries. Note that the filename from the MacPython ecosystem is deceptive to the true version number, but we have the hash in the filename & the ctypes version check in some CI entries.

The result on my fork, which I suspect we'll see here again as well--is that CI will be all-green for everything except the emulation case, which actually gets **worse** with the upstream patch on Skylake & produces 15 fails intead of 1 in linalg suite.

That probably means we need to consider another round of MacPython  openblas builds with recent OpenBLAS again--someone else should feel free to do that if they want, otherwise I'll get back to it at some point. https://github.com/xianyi/OpenBLAS/pull/2111 may help

Once there's a working patch for SDE CI entry, I'll probably have to replicate some of this effort downstream in SciPy before I can justify the OpenBLAS churn there as well.

martin-frbg commented:
It is unclear to me what you are doing here - is that report about 13401 still failing and an additional 15 failures based on the full OpenBLAS PR2111 as merged in current OpenBLAS-develop or on its earlier state with only one of the two SkylakeX xCOPY kernels disabled ?

tylerjereddy commented:
@martin-frbg It is an earlier state with only one of the kernels disabled, so your most recent fixes aren't included in the OpenBLAS used here. Some people had speculated that the earlier fix might be sufficient, but as you noted there was certainly no guarantee of that, and this demonstrates that indeed it was premature to assume that that fix was sufficient for NumPy.

tylerjereddy commented:
I've tried bumping OpenBLAS to latest develop branch commit again in https://github.com/MacPython/openblas-libs/pull/3 -- if that builds then I'll try again here later with the newer version

tylerjereddy committed:
TST, MAINT: bump Azure OpenBLAS

* another bump of Azure CI OpenBLAS version
as we attempt to incorporate appropriate patches
for SkylakeX-related linalg issues

tylerjereddy commented:
Testing on my fork / Azure suggests that the most recent MacPython ecosystem build of OpenBLAS is now passing linalg tests, including the new one here, with NumPy under SkylakeX emulation. I've pushed the most recent OpenBLAS version bump to this PR for reviewer confirmation.

I think this means we may be in good shape for wheels for NumPy again, and probably for SciPy too (though I should likely replicate this effort downstream to be sure).

tylerjereddy commented:
Alright, I've lifted the draft status since Azure results look good with most recent OpenBLAS + SkylakeX emulation. Whether this belongs in NumPy proper is not a debate I'm super-interested in though--should go somewhere in the ecosystem.

For now, this allows me to move forward with other things--I've preserved the git history for the CI results demo of fail -> pass, for now anyway. They get purged for Azure in 30 days, I think.

charris commented:
This check is moving to OpenBLAS, so closing.

tylerjereddy commented:
The regression guard is now in place in upstream CI.
